bin/parallel.py

```diff
@@ -26,74 +25,119 @@ def __lt__(self, other):
             return self.id < other.id
 
 
-class Parallel:
-    # f(task): the function to run on each queue item.
-    # num_threads: True: the configured default
-    #              None/False/0: disable parallelization
-    def __init__(self, f, num_threads=True, pin=False):
+class AbstractQueue:
+    def __init__(self, f, pin):
         self.f = f
-        self.num_threads = config.args.jobs if num_threads is True else num_threads
-        self.pin = pin and not util.is_windows() and not util.is_bsd()
+        self.pin = pin
+        self.num_threads = 1
+
+        # min heap
+        self.tasks: list[QueueItem] = []
+        self.total_tasks = 0
+        self.missing = 0
+
+        self.aborted = False
 
         # mutex to lock parallel access
         self.mutex = threading.RLock()
+
+    def __enter__(self):
+        self.mutex.__enter__()
+
+    def __exit__(self, *args):
+        self.mutex.__exit__(*args)
+
+    # Add one task. Higher priority => done first
+    def put(self, task, priority=0):
+        raise "Abstract method"
+
+    # By default, do nothing on .join(). This is overridden in ParallelQueue.
+    def join(self):
+        return
+
+    def done(self):
+        raise "Abstract method"
+
+    def abort(self):
+        self.aborted = True
+
+
+class SequentialQueue(AbstractQueue):
+    def __init__(self, f, pin):
+        super().__init__(f, pin)
+
+    # Add one task. Higher priority => done first
+    def put(self, task, priority=0):
+        # no task will be handled after self.abort() so skip adding
+        if self.aborted:
+            return
+
+        self.total_tasks += 1
+        heapq.heappush(self.tasks, QueueItem(task, priority, self.total_tasks))
+
+    # Execute all tasks.
+    def done(self):
+        if self.pin:
```

Comment:
I don't think pinning really makes much sense for the sequential queue, so these `if`s here can just be dropped I'd say

Reply:
I do think this makes sense actually, because I want my timings to be consistent, especially in sequential mode :stuck_out_tongue: